New method ActivationFrequency::initializeToValue

ctrl-z-9000-times opened the pull request:
void initializeToValue( Real initialValue );

 Assign an initial activation frequency to all bits in the SDR.  If this method is used then this metric will always use the exponential moving average, and skip over the regular average which is usually applied to  the first "period" many samples.

This method is useful for using this metric as part of boosting algorithms which seek to push the activation frequencies to a target value. These algorithms will overreact to the default early behavior of this class during the first "period" many samples.

This is used by the ColumnPooler, PR #285.

breznak commented on src/nupic/utils/SdrMetrics.hpp:
```diff
@@ -200,6 +200,19 @@ class ActivationFrequency : public MetricsHelper_ {
+    /**
+     * Assign an initial activation frequency to all bits in the SDR.  If this
+     * method is used then this metric will always use the exponential moving
+     * average, and skip over the regular average which is usually applied to
```
I don't like the fact that the name suggests "initialize to a value", while it enforces a computation mode (EMA vs normal average). 
Can you make it an argument? `initialize(Real initValue, bool=useExponentialMovingAverage=false)` ?

ctrl-z-9000-times replied:
No, because the argument initialValue requires that it always use the exponential rolling average, and never use the regular average.  The regular average will discard the initial value.

breznak replied:
nit, then maybe name initializeExponentialMovingAvg()?

breznak commented on src/nupic/utils/SdrMetrics.hpp:
```diff
@@ -200,6 +200,19 @@ class ActivationFrequency : public MetricsHelper_ {
+     * This method is useful for using this metric as part of boosting
+     * algorithms which seek to push the activation frequencies to a target
+     * value. These algorithms will overreact to the default early behavior of
+     * this class during the first "period" many samples.
```
the user will have to remember to handle it, can you make it more obvious instead? Maybe returning NaN, or 0.5f for the first period samples?

ctrl-z-9000-times replied:
> Maybe returning NaN, or 0.5f for the first period samples?

Then the user would have to remember to skip the first period many samples.

ctrl-z-9000-times commented on bindings/py/tests/sdr/Metrics_test.py:
```diff
@@ -86,6 +86,16 @@ def testAF_Example(self):
+        AF.initializeToValue( .02 )
+        assert( np.all( AF.activationFrequency == .02 ))
+        X.randomize( 1 )
+        assert( np.all( AF.activationFrequency > .025 ))
```
I will make this test deterministic so that it never fails CI.

I would like to keep the range checks because: the exact value depends on an arbitrary random seed which is not what this is testing.

ctrl-z-9000-times replied:
Actually, the seed is not needed since this uses 100% sparsity which is not very random.

The reason I use a range check here is because it true value is not 3%, it's very close to 3% but since it uses an exponential rolling avg it isn't quite 3%, it's a little lower.  I will change this to use exact equality instead of the current range check.

ctrl-z-9000-times commented on src/nupic/utils/SdrMetrics.cpp:
```diff
@@ -132,10 +132,25 @@ ActivationFrequency::ActivationFrequency( SDR &dataSource, UInt period )
 
 void ActivationFrequency::initialize(UInt size) {
     activationFrequency_.assign( size, 1234.567f );
```
It is (supposed to be) obviously invalid data.  It's a placeholder value which the user will see if they check the results before giving this any data samples.

ctrl-z-9000-times commented on src/nupic/utils/SdrMetrics.cpp:
```diff
@@ -132,10 +132,25 @@ ActivationFrequency::ActivationFrequency( SDR &dataSource, UInt period )
+    NTA_ASSERT( samples_ < 1 ) << "ActivationFrequency already in use.";
+    NTA_CHECK( not alwaysExponential_ ) << "ActivationFrequency already initialize to a value.";
+    NTA_CHECK( initialValue >= 0.0f );
+    NTA_CHECK( initialValue <= 1.0f );
```
Actually, this is an activation frequency, not a permanence.  The valid range of activation frequencies is always a fraction in the range [0, 1].

ctrl-z-9000-times commented:
I changed this to be part of the constructor, as an optional default argument.  I also added the following line to the docs, which I think clarifies what this does:
"Makes this ActivationFrequency instance think that it is the result of a long running process (even though it was just created)."

breznak commented on src/nupic/utils/SdrMetrics.cpp:
```diff
@@ -117,32 +117,32 @@ std::ostream& operator<<(std::ostream& stream, const Sparsity &S)
-ActivationFrequency::ActivationFrequency( const vector<UInt> dimensions, UInt period )
+ActivationFrequency::ActivationFrequency( const vector<UInt> &dimensions,
+                                          UInt                period,
+                                          Real                initialValue )
```
nit, const, const

ctrl-z-9000-times replied:
Done.

I just changed a bunch of things in the SDR & Metrics to use `const &` where able and appropriate.

breznak commented on src/nupic/utils/SdrMetrics.hpp:
```diff
@@ -188,30 +188,30 @@ class ActivationFrequency : public MetricsHelper_ {
      * instance by assigning to this SDR.
      *
      * @param period Time scale for exponential moving average.
```
OT, but please explain time-scale even a bit more

ctrl-z-9000-times replied:
Would it help if I changed it from "time scale" to "time constant"?  Since the later appears to be the correct terminology and it even has its own wiki page: https://en.wikipedia.org/wiki/Time_constant

breznak commented on src/nupic/utils/SdrMetrics.hpp:
```diff
@@ -188,30 +188,30 @@ class ActivationFrequency : public MetricsHelper_ {
+     * "period" many samples.
+     * Note: This argument is useful for using this metric as part of boosting
+     *       algorithms which seek to push the activation frequencies to a
+     *       target value. These algorithms will overreact to the default early
```
overreact, or would, if initValue (EMA) were not used?

ctrl-z-9000-times commented:
Breznak,  I made a bunch of small changes, but no functional changes, just docs & minor code cleanup.  This is ready for another review when you have the time.  Thanks
